Introduction to npm

ckaia opened the pull request:
@marudy, Please review and provide feedback accordingly

ckaia committed:
Initial commit npm #298

ckaia committed:
Create wb1

ckaia committed:
Rename wb1 to wb1.md

ckaia committed:
Add npm definition, commands and scenarios

ckaia committed:
Enhance command and scenarios information #298

ckaia committed:
Add run scripts. Enhance defintion, command and scenarios #298

ckaia committed:
frename file wb1.md to npm1.md #289

ckaia committed:
Initial commit nvm #298

ckaia committed:
Adding nvm definition and commands #298

marudy commented on tools/npm/npm1.md:
```diff
@@ -0,0 +1,158 @@
+# Definition
```
let's find a better name :) like 'what-is-npm.md' or just 'npm' or 'npm-guide' or .... You get the point ;)

ckaia replied:
👍

marudy commented on tools/npm/npm1.md:
```diff
@@ -0,0 +1,158 @@
+# Definition
+
+### What is npm?
+**npm** is the **node package manager** that makes it easy for JavaScript developers to share the code that they've created to solve particular problems, and for other developers to reuse that code in their own applications.
```
I like that we start with what is npm but the description is a bit generic - let's try to put something more solid that is going to be super clear for a newbie :slightly_smiling_face:

ckaia replied:
:+1:

marudy commented on tools/npm/npm1.md:
```diff
@@ -0,0 +1,158 @@
+### What is npm?
+**npm** is the **node package manager** that makes it easy for JavaScript developers to share the code that they've created to solve particular problems, and for other developers to reuse that code in their own applications.
+
+### What are node package?
```
singular or plural !

ckaia replied:
:sunglasses:

marudy commented on tools/npm/npm1.md:
```diff
@@ -0,0 +1,158 @@
+### Installing **npm**
+**npm** is installed as part of the **node.js**
+
+Check **node** version 
```
node is less important in this tutorial, let's put this info below the command for npm version

ckaia replied:
👍

marudy commented on tools/npm/npm1.md:
````diff
@@ -0,0 +1,158 @@
+npm -v
+```
+
+### Initialialize npm
````
Initialize project (with the help of npm) !

ckaia replied:
👍

marudy commented on tools/npm/npm1.md:
````diff
@@ -0,0 +1,158 @@
+
+A process will be started that will create **package.json** file in your current folder. A questionary command line is initiated where at least **name** and **version** should be provided.
+```sh
+npm init
````
i think it's best to put the command first in this section

ckaia replied:
👍

marudy commented on tools/npm/npm1.md:
```diff
@@ -0,0 +1,158 @@
+
+### Install a package locally
+
+Install the **dependencies** and **devDependencies** dev read from **package.json** in the local **node_modules** folder.
```
Very good here, but let's take 1 sentence or 2 to explain what is dependencies and what is devDependencies

**EDIT:** I just saw that you do it below, i think this information must be shown earlier to the reader

marudy replied:
And no need to mention node_modules folder here i believe. You mention its creation very well below :slightly_smiling_face:

ckaia replied:
👍

marudy commented on tools/npm/npm1.md:
````diff
@@ -0,0 +1,158 @@
+npm install react-native-cli -g
+```
+**npm** will install package **react-native-cli** as a computer program. The package is stored in a global path on your system.
+> **-g** flag: global install flag, if used **react-native-cli** package can run from any path
````
this markdown is not very well visible in view mode when i read the file as a whole. Maybe try something different?

ckaia replied:
👍

marudy commented on tools/npm/npm1.md:
````diff
@@ -0,0 +1,158 @@
+
+Custom scripts run as
+```sh
+npm run deploy
````
maybe put run in bold to differentiate from previous standard scripts?

ckaia replied:
😃

marudy commented on tools/npm/npm1.md:
````diff
@@ -0,0 +1,158 @@
+npm run deploy
+```
+
+# Scenarios
````
Maybe rename it to something like 'Hands on examples?'

ckaia replied:
cool

marudy commented on tools/nvm/nvm1.md:
```diff
@@ -0,0 +1,27 @@
+# Definition
+
+### What is **nvm**?
+**nvm** is a **node version manager** and is used to upgrade node version [nvm GitHub](https://github.com/creationix/nvm). It is a simple bash script to manage multiple active **node.js** versions.
```
nvm **stands for** ...

marudy replied:
I cannot understand exactly the rest of the 1st sentence, please rephrase ! :slightly_smiling_face:

ckaia replied:
👍

marudy commented on tools/nvm/nvm1.md:
````diff
@@ -0,0 +1,27 @@
+nvm install <node-version>
+```
+
+### Make new version default
````
Change default version sounds a bit better :slightly_smiling_face:

ckaia replied:
:stuck_out_tongue:

marudy commented on tools/nvm/nvm1.md:
````diff
@@ -0,0 +1,27 @@
+nvm ls
+```
+
+Change version
````
This is repeated, so we don't need it !

marudy commented on tools/nvm/nvm1.md:
````diff
@@ -0,0 +1,27 @@
+nvm alias default <node-version>
+```
+
+### Change version
````
No need for this title, just keep the command below

ckaia replied:
👍

marudy commented:
@ckaia 

I had a look and i want to say that the information you have gathered and put is super!!! :smile: I have requested some changes regarding phrasing and sometimes content. Have a look and let's do a second iteration !

ckaia committed:
Applying changes on review feedback #298

npm definition extension
Scenarios section changed to Hands on activities
Ordering changes and small fixes

ckaia committed:
Renaming file npm1.md to npm-guide.md #298

ckaia committed:
Applying changes after review #298

Update definition
Combine Upgrade node version with Change version to Upgrade node version or change version

ckaia committed:
Renaming file nvm1.md to nvm-guide.md #298

ckaia commented:
@marudy, please check updated npm and nvm guide with updated content after your valueable feedback!

marudy committed:
Stuctured the file better and re-wrote some content

ckaia committed:
Merge pull request #2 from marudy/npm-patch-1

Restructuring & some content rewriting

ckaia committed:
Removing empty files README.md from npm and nvm. #298)

ckaia committed:
Adding package.json definition #298

Adding step 3 in "Install a node package".

ckaia commented:
@marudy, please review updated npm guide

marudy committed:
Small content rephrasing

marudy committed:
Minor changes

marudy commented:
@ckaia did some small changes. Have a look as well and we are good from my side.

I added @Knorcedger as a reviewer in case we get some more feedback.

ckaia committed:
Update Installing npm link.

Provide installation commands from terminal for linux users. #298

Knorcedger commented on tools/npm/npm-guide.md:
```diff
@@ -0,0 +1,183 @@
+
+### Where can i get the Node packages from?
+
+Users open source and publish their packages in [npm Registry](https://www.npmjs.com/search). Since these packages are open and free, developers usually check README, stats and stars of each package in [GitHub](https://www.github.com) in order to decide how useful is a library before using it.
```
And number of installs (stats are on the package page on npm registry)

ckaia replied:
I change it to the following:

> Developers open source and publish their packages in [npm Registry](https://www.npmjs.com/search). 

> Since these packages are open and free, users are recommended to consult the Stats at the right of the web page of npm Registry for a package. For example see Stats of [lodash](https://www.npmjs.com/package/lodash) package. There is displayed the number of installments along with the GitHub link which provides the stars and README of each package in order to decide how useful is a library before using it.

Knorcedger commented on tools/npm/npm-guide.md:
```diff
@@ -0,0 +1,183 @@
+
+npm is installed along with [NodeJS](https://nodejs.org/en/download/) installation. This link provides NodeJS binaries and installers.
+
+Alternatively for linux users run the following command from terminal replacing the node version `setup_8.x`. This example is for node version 8.
```
This is not the best way to do it. It's a lot to explain, please contact me to talk about it

Knorcedger commented on tools/npm/npm-guide.md:
````diff
@@ -0,0 +1,183 @@
+}
+```
+
+It holds the list of all dependencies and devDependencies packages. Dependencies are the modules needed to be installed for a project to run while devDependencies are the modules used by the developer to make his life easier. 
````
Let's add a link to the official npm docs for package.jaon for more info.

ckaia replied:
:+1:

Knorcedger commented on tools/npm/npm-guide.md:
````diff
@@ -0,0 +1,183 @@
+npm install <package_name>
+```
+
+Install single package globally. That bein said, packages installed that way be run from any path.
````
We need to explain that this should generally be avoided and only utils packages should be installed globally.

ckaia replied:
:+1:

Knorcedger commented on tools/npm/npm-guide.md:
````diff
@@ -0,0 +1,183 @@
+npm list
+```
+
+Print to stdout only the dependencies of the first level. This command is usefull to check the versions of the packages installed locally.
````
Explain what first level is plz (packages listed in package.json)

ckaia replied:
:+1:

Knorcedger commented on tools/nvm/nvm-guide.md:
````diff
@@ -0,0 +1,24 @@
+### Install a new version
+
+```sh
+nvm install <node-version>
````
I think this is NVM install-remote or something. I might be wrong and have no way of testing it right now. Can we plz verify it?

ckaia replied:
As far as it concerns the installation of a node version using nvm, the command according to nvm README is [nvm install \<node-version\>](https://github.com/creationix/nvm#usage).

There is a command [nvm ls-remote](https://github.com/creationix/nvm#listing-versions) that provides the versions available for installment.

In [How To Install Node.js on Ubuntu 16.04](https://www.digitalocean.com/community/tutorials/how-to-install-node-js-on-ubuntu-16-04) it uses `nvm ls-remote` to check the available versions and then `npm install <node-version>` to install a specific version.

Knorcedger reviewed (changes requested):
We are missing instructions on how to install specific versions of a package. Also, add more links to official docs on each sections for people to read more. Generally, very well done!

lampropoi commented on tools/npm/npm-guide.md:
````diff
@@ -0,0 +1,183 @@
+npm init
+```
+
+### Installation
````
I believe that is good to also mention that if we want to install a package which will be used only during the development, this should be added in devDependencies.

marudy replied:
it's mentioned in lines 49-52 right ?

Knorcedger replied:
Ioanna is right, by default npm 5 assumes --save and saves to dependencies, but if we want something saved as a devDependency, we must use `npm install reqlog --save-dev`. We have to mention this difference and this command as an example.

ckaia replied:
:+1:

ckaia committed:
Adding updates after review #298

Adding further clarifications where needed.
- install globally
- npm list with depth 0
- Stas in npm package registry

ckaia committed:
Changes upon review #298

Adding how to install specific package versions.
Adding links to official docs where needed.
Adding install package in devDependencies.

ckaia committed:
Changes upon review #298

Adding links reference where needed

ckaia committed:
Updating on review PR#1, issue #298

Changing installing npm command
